Servers/DHCPClient/DHCPv4Client.cpp

```diff
@@ -0,0 +1,280 @@
+    sockaddr_in dst;
+    memset(&dst, 0, sizeof(dst));
+    dst.sin_family = AF_INET;
+    dst.sin_port = htons(67);
+    dst.sin_addr.s_addr = IPv4Address { 255, 255, 255, 255 }.to_u32();
+    memset(&dst.sin_zero, 0, sizeof(dst.sin_zero));
+
+    auto rc = sendto(fd, &packet, sizeof(packet), 0, (sockaddr*)&dst, sizeof(dst));
+    if (rc < 0) {
+        dbg() << "sendto failed with " << strerror(errno);
+        // FIXME: what do we do here?
+    }
+}
+
+static void set_params(const InterfaceDescriptor& iface, const IPv4Address& ipv4_addr, const IPv4Address& netmask, const IPv4Address& gateway)
+{
+    int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+    if (fd < 0) {
+        dbg() << "ERROR: socket";
+        return;
+    }
+
+    struct ifreq ifr;
+    memset(&ifr, 0, sizeof(ifr));
+    strncpy(ifr.ifr_name, iface.m_ifname.characters(), IFNAMSIZ);
+
+    {
+        ifr.ifr_addr.sa_family = AF_INET;
+        ((sockaddr_in&)ifr.ifr_addr).sin_addr.s_addr = ipv4_addr.to_in_addr_t();
+
+        int rc = ioctl(fd, SIOCSIFADDR, &ifr);
+        if (rc < 0) {
+            dbg() << "ERROR: ioctl(SIOCSIFADDR)";
+        }
+    }
+    {
+        ifr.ifr_addr.sa_family = AF_INET;
+        ((sockaddr_in&)ifr.ifr_netmask).sin_addr.s_addr = netmask.to_in_addr_t();
+
+        int rc = ioctl(fd, SIOCSIFNETMASK, &ifr);
+        if (rc < 0) {
+            dbg() << "ERROR: ioctl(SIOCSIFADDR)";
+        }
+    }
+    {
+        struct rtentry rt;
+        memset(&rt, 0, sizeof(rt));
+
+        rt.rt_dev = const_cast<char*>(iface.m_ifname.characters());
+        rt.rt_gateway.sa_family = AF_INET;
+        ((sockaddr_in&)rt.rt_gateway).sin_addr.s_addr = gateway.to_in_addr_t();
+        rt.rt_flags = RTF_UP | RTF_GATEWAY;
+
+        int rc = ioctl(fd, SIOCADDRT, &rt);
+        if (rc < 0) {
+            dbg() << "Error: ioctl(SIOCADDRT)";
+        }
+    }
+}
+
+DHCPv4Client::DHCPv4Client(Vector<InterfaceDescriptor> ifnames)
+    : m_ifnames(ifnames)
+{
+    m_server = Core::UDPServer::construct(this);
+    m_server->on_ready_to_receive = [this] {
+        auto buffer = m_server->receive(sizeof(DHCPv4Packet));
+        dbg() << "Received " << buffer.size() << " bytes";
+        auto& packet = *(DHCPv4Packet*)buffer.data();
+        process_incoming(packet);
+    };
+
+    if (!m_server->bind({}, 68))
+        dbg() << "We were already bound???";
+
+    for (auto& iface : m_ifnames)
+        dhcp_discover(iface);
+}
+
+DHCPv4Client::~DHCPv4Client()
+{
+}
+
+void DHCPv4Client::handle_offer(const DHCPv4Packet& packet, const ParsedDHCPv4Options& options)
+{
+    dbg() << "We were offered " << packet.yiaddr().to_string() << " for " << options.get<u32>(DHCPOption::IPAddressLeaseTime).value_or(0);
+    auto opt_transaction = m_ongoing_transactions.get(packet.xid());
+    if (!opt_transaction.has_value()) {
+        dbg() << "we're not looking for " << packet.xid();
+        return;
+    }
+    auto* transaction = const_cast<DHCPv4Transaction*>(opt_transaction.value());
```

Comment:
You can simplify this with the use of `value_or`:

```suggestion
    auto* transaction = m_ongoing_transactions.get(packet.xid()).value_or(nullptr);
    if (!transaction) {
        dbg() << "we're not looking for " << packet.xid();
        return;
    }
```